builtin/logical/database/credentials.go

```diff
@@ -170,3 +175,188 @@ func (kg rsaKeyGenerator) configMap() (map[string]interface{}, error) {
 	}
 	return config, nil
 }
+
+type ClientCertificateGenerator struct {
+	CommonNameTemplate string `mapstructure:"common_name_template,omitempty"`
+	CAPrivateKey       string `mapstructure:"ca_private_key,omitempty"`
+	CACert             string `mapstructure:"ca_cert,omitempty"`
+	KeyType            string `mapstructure:"key_type,omitempty"`
+	KeyBits            int    `mapstructure:"key_bits,omitempty"`
+	SignatureBits      int    `mapstructure:"signature_bits,omitempty"`
+
+	parsedCABundle *certutil.ParsedCertBundle
+	cnProducer     template.StringTemplate
+}
+
+// newClientCertificateGenerator returns a new ClientCertificateGenerator
+// using the given config. Default values will be set on the returned
+// ClientCertificateGenerator if not provided in the config.
+func newClientCertificateGenerator(config map[string]interface{}) (ClientCertificateGenerator, error) {
+	var cg ClientCertificateGenerator
+	if err := mapstructure.WeakDecode(config, &cg); err != nil {
+		return cg, err
+	}
+
+	switch cg.KeyType {
+	case "rsa":
+		switch cg.KeyBits {
+		case 0, 2048, 3072, 4096:
+		default:
+			return cg, fmt.Errorf("invalid key_bits")
+		}
+	case "ec":
+		switch cg.KeyBits {
+		case 0, 224, 256, 384, 521:
+		default:
+			return cg, fmt.Errorf("invalid key_bits")
+		}
+	case "ed25519":
+		// key_bits ignored
+	default:
+		return cg, fmt.Errorf("invalid key_type")
+	}
+
+	switch cg.SignatureBits {
+	case 0, 256, 384, 512:
```

Comment:
Just since I saw it explicitly listed in the example on the call, I'd default to `0` for auto-detection here in your API. Nobody should care about this parameter much, outside of very few users, and I'd even perhaps suggest eliding it altogether until someone explicitly requests it tbh.

Reply:
This is what we're doing right now for setting the default signature bits when 0 is provided via the API: https://github.com/hashicorp/vault/blob/328929264668bacf82d6802ac5b456cc5f5c6672/builtin/logical/database/credentials.go#L284-L307. Could you advise on if this look correct or if there was some other way to auto-detect that you had in mind? Thanks!

Reply:
Aha, I was thinking this function: `ValidateDefaultOrValueKeyTypeSignatureLength`

Reply:
Something makes me think you could pass through `0` as signatureBits fwiw. Not tried it recently though :-D

Reply:
Thanks! It does look like you can. We'll take a look to see if we can use this.